Source/NSButtonCell.m

```diff
@@ -1885,38 +1880,49 @@ - (id) initWithCoder: (NSCoder*)aDecoder
   else
     {
       BOOL tmp;
+      uint32_t tmp2;
       int version = [aDecoder versionForClassName: @"NSButtonCell"];
       NSString *key = nil;
 
       [aDecoder decodeValueOfObjCType: @encode(id) at: &key];
-      [self setKeyEquivalent: key]; // Set the key equivalent...
+      // Hack to correct a Gorm problem, there "\n" is used instead of "\r".
+      if ([key isEqualToString: @"\n" ])
+        {
+          key = @"\r";
+        }
+      [self setKeyEquivalent: key];
 
       [aDecoder decodeValueOfObjCType: @encode(id) at: &_keyEquivalentFont];
       [aDecoder decodeValueOfObjCType: @encode(id) at: &_altContents];
       [aDecoder decodeValueOfObjCType: @encode(id) at: &_altImage];
       [aDecoder decodeValueOfObjCType: @encode(BOOL) at: &tmp];
       _buttoncell_is_transparent = tmp;
-      [aDecoder decodeValueOfObjCType: @encode(unsigned int)
-                                   at: &_keyEquivalentModifierMask];
+      [aDecoder decodeValueOfObjCType: @encode(uint32_t)
+                                   at: &tmp2];
+      _keyEquivalentModifierMask = (NSUInteger)tmp2;
       if (version <= 2)
         {
           _keyEquivalentModifierMask = _keyEquivalentModifierMask << 16;
         }
-      [aDecoder decodeValueOfObjCType: @encode(unsigned int)
-                                   at: &_highlightsByMask];
-      [aDecoder decodeValueOfObjCType: @encode(unsigned int)
-                                   at: &_showAltStateMask];
+      [aDecoder decodeValueOfObjCType: @encode(uint32_t)
+                                   at: &tmp2];
+      _highlightsByMask = (NSInteger)tmp2;
+      [aDecoder decodeValueOfObjCType: @encode(uint32_t)
+                                   at: &tmp2];
+      _showAltStateMask = (NSInteger)tmp2;
 
       if (version >= 2)
         {
           [aDecoder decodeValueOfObjCType: @encode(id) at: &_sound];
           [aDecoder decodeValueOfObjCType: @encode(id) at: &_backgroundColor];
           [aDecoder decodeValueOfObjCType: @encode(float) at: &_delayInterval];
           [aDecoder decodeValueOfObjCType: @encode(float) at: &_repeatInterval];
-          [aDecoder decodeValueOfObjCType: @encode(unsigned int)
-                                       at: &_bezel_style];
-          [aDecoder decodeValueOfObjCType: @encode(unsigned int)
-                                       at: &_gradient_type];
+          [aDecoder decodeValueOfObjCType: @encode(uint32_t)
```

Comment:
I believe this is the right thing to do.  Going from "unsigned int" to uint32_t is fine as long as we can still read current archives.

Reply:
I will try to express my thoughts in a more concrete way.   Firstly, I think in general the proper way to do things is encode/decode the type of the actual instance variable (except in cases where we are encoding/decoding information that is not stored in a single instance variable of course).  Secondly, while the standard API supports that perfectly well, it does not provide for compile-time type checking, depending on the author of the code to supply correct arguments, and this is clearly a disadvantage.  To provide a mechanism to take advantage of compile-time type checking, I'm suggesting private stuff (eg added to GSPrivate.h in the base library) as follows:

/** Macro to create static wrapper functions to enable the compiler
 * to warn about incorrect variable types during decoding.
 */
#define GS_DECODER(name,type) \
static inline void name(NSCoder *coder, type *value) \
{ \
  [coder decodeValueOfObjCType: @encode(type) at: value]; \
} 

/** Macro to create static wrapper functions to enable the compiler
 * to warn about incorrect variable types during decoding.
 */
#define GS_ENCODER(name,type) \
static inline void name(NSCoder *coder, type *value) \
{ \
  [coder encodeValueOfObjCType: @encode(type) at: value]; \
} 

Using those macros we can then create inline function definitions which are local to the files where they are used (so no pollution of the public API/ABI) and which should have zero runtime cost.

eg.

Somewhere before we want to use it, we do
  GS_ENCODER(encodeNSInteger,NSInteger)
which generates the definition of a static function:
  static inline void encodeNSInteger(NSCoder *coder, NSInteger *value)
  {
    [coder encodeValueOfObjCType: @encode(NSInteger) at: value]; 
  }

Then if we are encoding a class which has an instance variable

  unsigned myIvar;

the code to encode it would look like this:

  encodeNSInteger(aCoder, &myIvar);

The compiler would convert that inline function to 
  [coder encodeValueOfObjCType: @encode(NSInteger) at: &myValue];
so there's no runtime overhead of doing this.
But ... the compiler would do type checking of the argument to the inline function, and would generate a warning that a pointer to an unsigned integer is being passed to a function which expects a pointer to an NSInteger

Reply:
Actually, I just remembered we could use the ## to make a better (simpler) macro.
eg.
#define GS_ENCODER(type) \
static inline void encode_##type(NSCoder *coder, type *value) \
{ \
  [coder encodeValueOfObjCType: @encode(type) at: value]; \
}

so

GS_ENCODER(char)

would expand to

static inline void encode_char(NSCoder *coder, char *value)
{
  [coder encodeValueOfObjCType: @encode(char) at: value];
}

Reply:
I like the idea with the macros. This enables us to have it in gui independent of base. That way we don't need a base release before being able to use it in gui, as Riccardo's solution would require.
I will check whether this results in my compiler (gcc 10) giving proper warnings.

Reply:
I wrote:
In an ideal world, we could minimise errors encoding a variable foo by doing
[aCoder encodeValueOfObjCType: @encode(typeof(foo)) at &foo];
However I'm pretty sure the compilers don't support applying @encode() to typeof().

However, in looking at existing code I see we are already using that technique in some places in the base library, so it presumably *does* work for all compilers we target.  That really makes the idea of having extra wrappers to get the compiler to do type checking semi-irrelevant because using typeof() guarantees that we encode/decode correctly as the compiler will supply the type information.
What it doesn't do is handle the situation where we change a variable from signed to unsigned or vice-versa, since the compiler would automatically change the encoding to match, but archives generated by the old code would not be compatible with those generated by the new code, and we would not notice that the archive version should be changed until/unless we try to decode an old archive generated before we made the code change.

Reply:
I checkt with my compiler and the warnings I get from misusing your Marcos are great. It will get hard to use these incorrectly.
The second more compact version of the macros won't work for types like "unsigned int" or "long long". As that will result in a white space in the function name. But as long as we stick to simple type names we are good to go.
For gui I would put these macros into the file GSGuiPrivate.h.

In the decoding of NSButtonCell I would still use a temporary variable for four cases. Two where we need a signed integer, while an unsigned is encoded in Gorm files. And the other two where enums get decoded. This is something we could further discuss about.

Reply:
@rfm, I concur with Fred on this.  I like the macros you've written.  If they provide is some protection against this type of thing occurring in the future then it's worth doing.